File: brain4it-manager/brain4it-manager-web/src/main/webapp/js/dashboard/widgets/GaugeWidget.js
/* GaugeWidget */

Brain4it.GaugeWidget = function(dashboard, element)
{
  Brain4it.Widget.call(this, dashboard, element);
  this.label = null;
  this.getValuefunction = null;
  this.min = 0;
  this.max = 100;
  this.divisions = 10;
  this.decimals = 0;
  this.value = 0;
  this.remoteValue = 0;
  this.timerId = 0;
};

Brain4it.GaugeWidget.prototype = Object.create(Brain4it.Widget.prototype);

Brain4it.GaugeWidget.prototype.init = function(name, setup)
{
  var scope = this;
  
  this._onRemoteChange = this.onRemoteChange.bind(this);
  
  this.canvasElem = document.createElement("canvas");
  this.element.appendChild(this.canvasElem);
  var func = setup.getByName("get-value");
  if (func instanceof Brain4it.Reference)
  {
    this.getValueFunction = func.name;
    this.dashboard.monitor.watch(this.getValueFunction, this._onRemoteChange);
  }
  var value = setup.getByName("label");
  if (value)
  {
    this.label = String(value);
  }
  value = setup.getByName("min");
  if (Brain4it.isNumber(value))
  {
    this.min = Number(value);
  }
  value = setup.getByName("max");
  if (Brain4it.isNumber(value))
  {
    this.max = Number(value);
  }
  if (this.min >= this.max) 
    throw "max must be greater than min!";

  var value = setup.getByName("divisions");
  if (Brain4it.isNumber(value))
  {
    this.divisions = Number(value);
  }
  if (this.divisions < 5)
    throw "divisions must be greater than 4";    

  value = setup.getByName("decimals");
  if (Brain4it.isNumber(value))
  {
    this.decimals = Number(value);
  }
  if (this.decimals < 0)
    throw "decimals must be positive or zero";    

  var scope = this;
  window.addEventListener("resize",
    function() { scope.updateLayout(); }, false);
  this.updateLayout();
};

Brain4it.GaugeWidget.prototype.updateLayout = function()
{
  var width = this.element.clientWidth;
  var height = this.element.clientHeight;
  var pixelRatio = window.devicePixelRatio;
  this.canvasElem.width = width * pixelRatio;
  this.canvasElem.height = height * pixelRatio;
  this.canvasElem.style.width = width + "px";
  this.canvasElem.style.height = height + "px";
  this.paint();
};

Brain4it.GaugeWidget.prototype.paint = function()
{
  var width = this.canvasElem.width;
  var height = this.canvasElem.height;
  var size = Math.min(width, height);
  var margin = size / 10;
  var pointRadius = Math.round(size / 20);
  size -=  margin;
  var cx = width / 2;
  var cy = height / 2;
  var gaugeRadius = Math.round(size / 2);
  var pixelRatio = window.devicePixelRatio;

  var ctx = this.canvasElem.getContext("2d");
  ctx.clearRect(0, 0, width, height);

  // circle
  ctx.lineWidth = 2 * pixelRatio;
  ctx.beginPath();
  ctx.arc(cy, cy, gaugeRadius, 0, 2 * Math.PI);
  ctx.stroke();

  // ball
  ctx.lineWidth = pixelRatio;
  ctx.beginPath();
  ctx.arc(cx, cy, pointRadius, 0, 2 * Math.PI);
  ctx.stroke();

  var fontSize = Math.round(size / 15);
  ctx.font = "" + fontSize + "px Arial";

  var step = (this.max - this.min) / this.divisions;
  var radius1 = size / 2;
  var radius2 = radius1 - 0.1 * radius1;
  var radius3 = radius1 - 0.2 * radius1;

  var angle = 225;
  var stepAngle = 270.0 / this.divisions;

  for (var d = 0; d <= this.divisions; d++)
  {
    var i = Math.round(this.min + (d * step));
    var radians = angle * Math.PI / 180;
    var cosAngle = Math.cos(radians);
    var sinAngle = Math.sin(radians);

    var px = cx + radius1 * cosAngle;
    var py = cy - radius1 * sinAngle;
    var qx = cx + radius2 * cosAngle;
    var qy = cy - radius2 * sinAngle;
    var tx = cx + radius3 * cosAngle;
    var ty = cy - radius3 * sinAngle;

    ctx.beginPath();
    ctx.moveTo(px, py);
    ctx.lineTo(qx, qy);
    ctx.stroke();

    var valueLabel = String(i);

    var textWidth = ctx.measureText(valueLabel).width;

    var ox = textWidth / 2;

    ctx.fillText(valueLabel, tx - ox, ty + 0.3 * fontSize);

    angle -= stepAngle;
  }
  
  var div = Math.pow(10, this.decimals);
  var roundedValue = Math.round(this.remoteValue * div) / div;
  var valueString = String(roundedValue);

  var textWidth = ctx.measureText(valueString).width;
  var tx = cx - 0.5 * textWidth;
  var ty = cy + radius3 - fontSize;
  ctx.fillText(valueString, tx, ty);
  
  if (this.label !== null)
  {
    textWidth = ctx.measureText(this.label).width;
    tx = cx - 0.5 * textWidth;
    ty = cy + radius3 + 0.3 * fontSize;
    ctx.fillText(this.label, tx, ty);
  }

  var valueRatio = (this.value - this.min) / (this.max - this.min);
  if (valueRatio > 1) valueRatio = 1;
  else if (valueRatio < 0) valueRatio = 0;

  var valueAngle = 225 - valueRatio * 270;
  var radians = valueAngle * Math.PI / 180;
  var vx = cx + radius3 * Math.cos(radians);
  var vy = cy - radius3 * Math.sin(radians);

  ctx.lineWidth = 2 * pixelRatio;
  ctx.beginPath();
  ctx.moveTo(cx, cy);
  ctx.lineTo(vx, vy);
  ctx.stroke();
};

Brain4it.GaugeWidget.prototype.onRemoteChange = 
  function(functionName, value, serverTime)
{
  if (typeof value === "number")
  {
    this.remoteValue = value;
    this.animateGauge();
  }
};

Brain4it.GaugeWidget.prototype.animateGauge = function()
{
  if (this.timerId)
  {
    clearTimeout(this.timerId);
  }

  if (this.value !== this.remoteValue)
  {
    var dif = Math.abs(this.remoteValue - this.value);
    var variation = (this.max - this.min) / 100;
    if (dif < variation)
    {
      this.value = this.remoteValue;
    }
    else
    {
      this.value = this.value < this.remoteValue ? 
        this.value + variation : this.value - variation;
      var scope = this;
      this.timerId = setTimeout(function(){ scope.animateGauge(); }, 20);
    }
    this.paint();
  }  
};

Brain4it.Dashboard.prototype.widgetTypes['gauge'] = Brain4it.GaugeWidget;

File: brain4it-manager/brain4it-manager-android/src/main/java/org/brain4it/manager/android/view/GaugeWidget.java
/*
 * Brain4it
 * 
 * Copyright (C) 2018, Ajuntament de Sant Feliu de Llobregat
 * 
 * This program is licensed and may be used, modified and redistributed under 
 * the terms of the European Public License (EUPL), either version 1.1 or (at 
 * your option) any later version as soon as they are approved by the European 
 * Commission.
 * 
 * Alternatively, you may redistribute and/or modify this program under the 
 * terms of the GNU Lesser General Public License as published by the Free 
 * Software Foundation; either  version 3 of the License, or (at your option) 
 * any later version. 
 *   
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *   
 * See the licenses for the specific language governing permissions, limitations 
 * and more details.
 *   
 * You should have received a copy of the EUPL1.1 and the LGPLv3 licenses along 
 * with this program; if not, you may find them at: 
 *   
 *   https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *   http://www.gnu.org/licenses/ 
 *   and 
 *   https://www.gnu.org/licenses/lgpl.txt
 */
package org.brain4it.manager.android.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Color;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.view.View;
import java.util.TimerTask;
import org.brain4it.client.Monitor;
import org.brain4it.lang.BSoftReference;
import org.brain4it.lang.BList;
import org.brain4it.manager.android.DashboardActivity;
import org.brain4it.manager.android.DashboardWidget;
import org.brain4it.manager.widgets.GaugeWidgetType;
import org.brain4it.manager.widgets.WidgetType;

/**
 *
 * @author realor
 */
public class GaugeWidget extends View implements DashboardWidget
{
  protected DashboardActivity dashboard;
  protected String label;
  protected String getValueFunction;
  protected int min;
  protected int max;
  protected int divisions;
  protected int decimals;
  protected double value;
  protected double remoteValue;
  protected TimerTask timerTask;
  protected Paint linePaint1;
  protected Paint linePaint2;
  protected Paint textPaint;
  protected final Monitor.Listener monitorListener = new Monitor.Listener()
  {
    @Override
    public void onChange(String reference, final Object value, 
      long serverTime)
    {
      if (value instanceof Number)
      {
        remoteValue = ((Number)value).doubleValue();
        animateGauge();
      }
    }
  };

  public GaugeWidget(Context context)
  {
    this(context, null);
  }

  public GaugeWidget(Context context, AttributeSet attrs)
  {
    super(context, attrs);

    float density = context.getResources().getDisplayMetrics().density;    
    
    linePaint1 = new Paint();
    linePaint1.setColor(Color.BLACK);
    linePaint1.setStyle(Paint.Style.STROKE);
    linePaint1.setFlags(Paint.ANTI_ALIAS_FLAG);
    linePaint1.setStrokeWidth(density);

    linePaint2 = new Paint();
    linePaint2.setColor(Color.BLACK);
    linePaint2.setStyle(Paint.Style.STROKE);
    linePaint2.setFlags(Paint.ANTI_ALIAS_FLAG);
    linePaint2.setStrokeWidth(2 * density);

    textPaint = new Paint();
    textPaint.setColor(Color.BLACK);
    textPaint.setStyle(Paint.Style.FILL_AND_STROKE);    
    textPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
  }
  
  @Override
  public void init(DashboardActivity dashboard, String name, BList properties)
    throws Exception
  {
    this.dashboard = dashboard;
    
    GaugeWidgetType type = 
      (GaugeWidgetType)WidgetType.getType(WidgetType.GAUGE);
    
    type.validate(properties);
    
    label = type.getLabel(properties);
    min = type.getMin(properties);
    max = type.getMax(properties);
    divisions = type.getDivisions(properties);
    decimals = type.getDecimals(properties);
    
    BSoftReference func = type.getGetValueFunction(properties);
    if (func != null)
    {
      getValueFunction = func.getName();
      if (dashboard != null)
      {
        Monitor monitor = dashboard.getMonitor();
        monitor.watch(getValueFunction, monitorListener);
      }
    }
  }

  @Override
  public void onDraw(Canvas canvas)
  {
    int width = canvas.getWidth();
    int height = canvas.getHeight();
    int size = Math.min(width, height);
    int margin = size / 10;
    int pointRadius = size / 20;
    size -=  margin;
    int cx = width / 2;
    int cy = height / 2;
    float gaugeRadius = size / 2;

//    float paintWidth = Math.max(2f, (float)size / 400f);
//    linePaint1.setStrokeWidth(paintWidth);
//    linePaint2.setStrokeWidth(3 * paintWidth);
    textPaint.setTextSize(size / 15);
    
    // circle    
    canvas.drawCircle(cy, cy, gaugeRadius, linePaint2);
   
    // ball
    canvas.drawCircle(cx, cy, pointRadius, linePaint1);
    
    double step = (double)(max - min) / divisions;
    double radius1 = size / 2;
    double radius2 = radius1 - 0.1 * radius1;
    double radius3 = radius1 - 0.2 * radius1;

    double angle = 225;
    double stepAngle = 270.0 / divisions;
    Rect bounds = new Rect();
    
    for (int d = 0; d <= divisions; d++)
    {
      int i = (int)(min + (d * step));
      double radians = Math.toRadians(angle);
      double cosAngle = Math.cos(radians);
      double sinAngle = Math.sin(radians);
      
      int px = cx + (int)(radius1 * cosAngle);
      int py = cy - (int)(radius1 * sinAngle);
      int qx = cx + (int)(radius2 * cosAngle);
      int qy = cy - (int)(radius2 * sinAngle);
      int tx = cx + (int)(radius3 * cosAngle);
      int ty = cy - (int)(radius3 * sinAngle);
 
      canvas.drawLine(px, py, qx, qy, linePaint1);
      
      String valueLabel = String.valueOf(i);
      
      textPaint.getTextBounds(valueLabel, 0, valueLabel.length(), bounds);

      int ox = (int)(bounds.width() / 2);
      
      canvas.drawText(valueLabel, tx - ox, 
        (float)(ty + 0.8 * 0.5 * bounds.height()), textPaint);
      
      angle -= stepAngle;
    }
    
    double div = Math.pow(10, decimals);
    double roundedValue = Math.round(remoteValue * div) / div;
    String valueString;
    if (decimals == 0)
    {
      valueString = String.valueOf((int)roundedValue);
    }
    else 
    {
      valueString = String.valueOf(roundedValue);
    }
    textPaint.getTextBounds(valueString, 0, valueString.length(), bounds);    
    int tx = (int)(cx - (0.5 * bounds.width()));
    int ty = (int)(cy + radius3 - 1.5 * bounds.height());
    canvas.drawText(valueString, tx, ty, textPaint); 

    if (label != null)
    {
      textPaint.getTextBounds(label, 0, label.length(), bounds);
      tx = (int)(cx - (0.5 * bounds.width()));
      ty = (int)(cy + radius3 + 0.8 * 0.5 * bounds.height());
      canvas.drawText(label, tx, ty, textPaint); 
    }
    
    double valueRatio = (value - min) / (max - min);
    if (valueRatio > 1) valueRatio = 1;
    else if (valueRatio < 0) valueRatio = 0;
    
    double valueAngle = 225 - valueRatio * 270;
    double radians = Math.toRadians(valueAngle);
    int vx = cx + (int)(radius3 * Math.cos(radians));
    int vy = cy - (int)(radius3 * Math.sin(radians));
    
    canvas.drawLine(cx, cy, vx, vy, linePaint2);
  }
  
  private void animateGauge()
  {
    if (timerTask != null)
    {
      timerTask.cancel();
    }

    if (value != remoteValue)
    {
      double dif = Math.abs(remoteValue - value);
      double variation = (max - min) / 100.0;
      if (dif < variation)
      {
        value = remoteValue;
      }
      else
      {
        value = value < remoteValue ? 
          value + variation : value - variation;
        timerTask = new TimerTask()
        {
          @Override
          public void run()
          {
            animateGauge();
          }
        };
        if (dashboard != null)
        {
          dashboard.getTimer().schedule(timerTask, 20);
        }
      }
      postInvalidate();
    }
  }
}
